clients/include/rocblas_arguments.hpp

```diff
@@ -324,4 +328,216 @@ static_assert(std::is_trivial<Arguments>{},
+    e_side,
+    e_uplo,
+    e_diag,
+    e_batch_count,
+    e_stride_a,
+    e_stride_b,
+    e_stride_c,
+    e_stride_d,
+    e_stride_x,
+    e_stride_y,
+    e_algo,
+    e_solution_index,
+
+};
+
+#define CTOKEN(n) const char* const c_tok_##n = #n
+CTOKEN(M);
+CTOKEN(N);
+CTOKEN(K);
+CTOKEN(lda);
+CTOKEN(ldb);
+CTOKEN(ldc);
+CTOKEN(ldd);
+CTOKEN(a_type);
+CTOKEN(b_type);
+CTOKEN(c_type);
+CTOKEN(d_type);
+CTOKEN(compute_type);
+CTOKEN(incx);
+CTOKEN(incy);
+CTOKEN(incb);
+CTOKEN(incd);
+CTOKEN(alpha);
+CTOKEN(beta);
+CTOKEN(transA);
+CTOKEN(transB);
+CTOKEN(side);
+CTOKEN(uplo);
+CTOKEN(diag);
+CTOKEN(batch_count);
+CTOKEN(stride_a);
+CTOKEN(stride_b);
+CTOKEN(stride_c);
+CTOKEN(stride_d);
+CTOKEN(stride_x);
+CTOKEN(stride_y);
+CTOKEN(algo);
+CTOKEN(solution_index);
+#undef CTOKEN
+
+class ArgumentModel
+{
+public:
+    ArgumentModel(const std::vector<arg_type>& params);
+    virtual ~ArgumentModel() {}
+
+    bool hasParam(arg_type a);
+
+    template <typename T>
+    static void print_value(std::ostream& str, const T& x)
+    {
+        Arguments::print_value(str, x);
+    }
+
+    // Float16 output
+    static void print_value(std::ostream& str, rocblas_half x)
+    {
+        Arguments::print_value(str, double(x));
+    }
+
+    template <typename T>
+    void log_args(std::ostream&    str,
+                  const Arguments& args,
+                  double           gpu_us,
+                  double           gflops,
+                  double           gpu_bytes = 0,
+                  double           cpu_us    = 0,
+                  double           norm1     = 0,
+                  double           norm2     = 0);
+
+    virtual void log_perf(std::stringstream& name_str,
+                          std::stringstream& val_str,
+                          const Arguments&   arg,
+                          double             gpu_us,
+                          double             gflops,
+                          double             gpu_bytes,
+                          double             cpu_us,
+                          double             norm1,
+                          double             norm2);
+
+protected:
+    std::vector<arg_type> m_args;
+};
+
+template <typename T>
+void ArgumentModel::log_args(std::ostream&    str,
+                             const Arguments& arg,
+                             double           gpu_us,
+                             double           gflops,
+                             double           gpu_bytes,
+                             double           cpu_us,
+                             double           norm1,
+                             double           norm2)
+{
+    std::stringstream name_list;
+    std::stringstream value_list;
+    const char        delim = ',';
+
+    auto print = [&](const char* const name, auto x) mutable {
+        name_list << name << delim;
+        print_value(value_list, x);
+        value_list << delim;
+    };
+
+#define CASE(x)                  \
```

Comment:
These macros and the `print` lambda are unnecessarily complicated. Just define a variadic template function and pass names and values:
```
print_values("M", M, 
             "N", N,
             "K", K, 
             );
```

There are examples in `logging.h`. I plan to generalize them to support different formats.

Reply:
The list of arguments is not known at compile time, only run time, thus the switch.   The lambda is a simple functor much like the one pre-existing in the base class, to me a variadic template is more complicated.

Reply:
> 
> 
> The list of arguments is not known at compile time, only run time, thus the switch. The lambda is a simple functor much like the one pre-existing in the base class, to me a variadic template is more complicated.

`constexpr` can be used to make the constants known at compile-time.